pkg/wallet/account/account.go

```diff
@@ -0,0 +1,265 @@
+	PublicKey    types.PublicKey           `yaml:"PublicKey,flow"`
+}
+
+func New(version *uint8, nickname string, address types.Address, salt [16]byte, nonce [12]byte, cipheredData types.EncryptedPrivateKey, publicKey types.PublicKey) (*Account, error) {
+	if version == nil {
+		version = new(uint8)
+		*version = AccountLastVersion
+	}
+
+	if *version > AccountLastVersion {
+		return nil, fmt.Errorf("unsupported account version: %d", *version)
+	}
+
+	if !NicknameIsValid(nickname) {
+		return nil, fmt.Errorf("invalid nickname: %s", nickname)
+	}
+
+	if err := address.Validate(address.Version, object.UserAddress, object.SmartContractAddress); err != nil {
+		return nil, fmt.Errorf("validating address: %w", err)
+	}
+
+	if err := cipheredData.Validate(cipheredData.Version, object.EncryptedPrivateKey); err != nil {
+		return nil, fmt.Errorf("validating encrypted private key: %w", err)
+	}
+
+	if err := publicKey.Validate(publicKey.Version, object.PublicKey); err != nil {
+		fmt.Println("52", publicKey.Version)
+		fmt.Println("53", publicKey.Kind)
+		return nil, fmt.Errorf("validating public key: %w", err)
+	}
+
+	return &Account{
+		Version:      version,
+		Nickname:     nickname,
+		Address:      address,
+		Salt:         salt,
+		Nonce:        nonce,
+		CipheredData: cipheredData,
+		PublicKey:    publicKey,
+	}, nil
+}
+
+func NewEmpty() Account {
+	return Account{
+		Version:  new(uint8),
+		Nickname: "",
+		Address: types.Address{
+			Object: &object.Object{
+				Kind:    object.UserAddress,
+				Version: 0x00,
+				Data:    nil,
+			},
+		},
+		Salt:  [16]byte{},
+		Nonce: [12]byte{},
+		CipheredData: types.EncryptedPrivateKey{
+			Object: &object.Object{
+				Kind:    object.EncryptedPrivateKey,
+				Version: 0x00,
+				Data:    nil,
+			},
+		},
+		PublicKey: types.PublicKey{
+			Object: &object.Object{
+				Kind:    object.PublicKey,
+				Version: 0x00,
+				Data:    nil,
+			},
+		},
+	}
+}
+
+// NewGenerated generates a new account with a random private key. It destroys the guarded password.
+func NewGenerated(guardedPassword *memguard.LockedBuffer, nickname string) (*Account, error) {
+	version := uint8(AccountLastVersion)
+
+	publicKeyBytes, privateKeyBytes, err := ed25519.GenerateKey(nil)
+	if err != nil {
+		return nil, fmt.Errorf("generating ed25519 key-pair: %w", err)
+	}
+
+	var salt [16]byte
+	_, err = rand.Read(salt[:])
+	if err != nil {
+		return nil, fmt.Errorf("generating random salt: %w", err)
+	}
+
+	var nonce [12]byte
+	_, err = rand.Read(nonce[:])
+	if err != nil {
+		return nil, fmt.Errorf("generating random nonce: %w", err)
+	}
+
+	aeadCipher, secretKey, err := crypto.NewSecretCipher(guardedPassword.Bytes(), salt[:])
+	guardedPassword.Destroy()
+	if err != nil {
+		return nil, fmt.Errorf("creating secret cipher: %w", err)
+	}
+
+	// secretBuffer is the secret key in clear without the version.
+	secretBuffer := memguard.NewBufferFromBytes(privateKeyBytes.Seed())
```

Comment:
You're mixing secret key and provate key. here you should talk about the private keyu instead.